### PR TITLE
Fix Memory release not correctly in Multipart Decoder

### PR DESCRIPTION
**Pull request by fredericBregier**
Motivation:
2 years ago a change remove the default clearing of all HttpData, whatever
they are disk based or memory based.

A lot of users were probably releasing HttpData directly, so there was no issue.
But now, it seems, and as the Javadoc said, that `decoder.destroy()` shall clean up
also Memory based HttpData, and not only Disk based HttpData as currently.

Change:
Add in `destroy()` method the necessary code to release if necessary
the underlying Memory based HttpDatas.

Change one Junit Test (using Mixed, Memory and Disk based factories)
in order to check the correctness of this behavior and to really act 
as a handler (releasing buffers or requests).

Modify one Junit core to check validity when a delimiter is present in the Chunk 
but not CRLF/LF (false delimiter), to ensure correctness.

Result:
No more issue on memory leak

Note that still the List and the Map are not cleaned, since they were not
before. No change is done on this, since it could produce backward issue compatibility.

Fix issues #11175 and #11184

### CODE REVIEW

**Comment by jameskleeh** [comment]
This does not fix #11174

**Comment by fredericBregier** [comment]
Have you tested it ? If so, then, I don't understand the issue correctly.
I mean what you report seems that data are not released correctly.

Does it concerned the "offered" chunks ? If so, then in Netty, the standard http codec handler does that for you. This is not the job of the MultipartDecoder.

So, I was believing that it was the allocated data from the MultipartDecoder. As it used the same Pool allocator than original chunk, they could be Direct Memory too. And therefore, giving your issue.
Can you explained more ?

**Comment by jameskleeh** [comment]
@fredericBregier I'll post a reply in my issue

**Comment by fredericBregier** [comment]
Notice: I think there is something wrong with this undecodedChunk when we end the offer part. It might have a refCnt > 1.
So I added a new commit for this. I know this shouldn't be, but that's what we already faced up in a previous cases and were hoping this simple fix with simple release was ok. Maybe not ? So I introduce this while loop.

**Comment by jameskleeh** [comment]
@fredericBregier I'll try exercising the app with this change and report back

**Comment by fredericBregier** [comment]
@jameskleeh Sorry, wrong commit again (too late, I need to sleep). If you clone it, redo it please.

**Comment by jameskleeh** [comment]
@fredericBregier This logic doesn't get executed in my example. The issue is the `release` in the `destroy` method. The first call to `release()` returns false and doesn't deallocate the memory. I verified calling it a second time resolves the issue

**Comment by fredericBregier** [comment]
@jameskleeh I guess I found the reason.
When the undecodedChunk is free up when max size is reached, it was badly released.
Moreover, when the first chunk is provided, instead of copying as for all other items, it is just retained. But as it is shared with the HttpRequest. So it will be released by HttpRequest releasing (or first HttpChunk release).
So I change to copy as other cases, and still check if undecodedChunk has to be released more than once (according to my test, it shall not, so we may remove the while loop for a unique release call).

Can you recheck ?

**Comment by fredericBregier** [comment]
@normanmaurer I agree, that's why I remove the second one in destroy. 
The first one, in offer, is necessary according to the previous code (the if that preceeds says refCnt > 1, so needing more than one release).
However, note that according to my test (and using no more retained for the first Chunk), even the first one (in offer) could be of no use (never goes within the else part).

**Comment by normanmaurer** [comment]
@fredericBregier not sure if there is just some miss-understanding. If there is any `while(!...release())` calls needed there is another bug which you just "hide".

**Comment by fredericBregier** [comment]
@normanmaurer 
I fully agree. That's why I first removed the second one (destroy), but let at that time the first one (offer).
The first one, we discussed about it some months ago and you propose this solution, but when I look again at that point, as there is a `if refCnt() ==1` , the `else` part is implicitely more than 1.
However, with the change made on allocation of undecodedChunk for first chunk, my guess is that this `else` case never occurs.
So I propose you to remove this while loop also.

2 possibilities:
1) remove the "if/else" and stay with the content of "if" (no release), assuming the refCnt > 1 never occurs
2) keep the "if/else" but with no while loop (only one release but strange since refCnt > 1)

WDYT?

**Comment by fredericBregier** [comment]
@normanmaurer I opt for the moment on the second solution (keeping your code), but I believe the `else` might be never used.
It is harmless...

**Comment by fredericBregier** [comment]
@normanmaurer I have completed the last commit with 2 things, in addition to remove all while loop
- in constructor, if the decoder raised an exception, one of the implementation was not doing a call to destroy
- fix Junit tests to act as within a handler where the http codec release the chunks or requests
It seems it was the reason of the memory leak detected within the CI/CD and probably related to issues too.

**Comment by fredericBregier** [comment]
@normanmaurer So the tests passed with no more Memory leak.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java [hunk 1]
```diff
@@ -179,12 +179,18 @@ public HttpPostMultipartRequestDecoder(HttpDataFactory factory, HttpRequest requ
...
-            parseBody();
+        try {
+            if (request instanceof HttpContent) {
+                // Offer automatically if the given request is als type of HttpContent
+                // See #1089
+                offer((HttpContent) request);
+            } else {
+                parseBody();
+            }
+        } catch (ErrorDataDecoderException e) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:190]: @fredericBregier can you do this change in a separate PR and also add a unit test ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java (hunk 1 above)
Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:190]: @normanmaurer I'm not sure I can. Let me explain.
First, this is already in place in HttpPostStandardRequestDecoder, so it should be also within Multipart too.

I notice this during PR since the deploy was wrong due to some Junit having a Leak detection error due to this missing.

So, in fact, there are already tests for that, but relying on leak detection step in CI/CD.

For Multipart, such as: 
- testMultipartRequestWithFileInvalidCharset
- testMultipartRequestWithFieldInvalidCharset
- testDecodeMalformedNotEncodedContentDispositionFieldParameters
- testDecodeMalformedBadCharsetContentDispositionFieldParameters

While on Standard one, we have:
- testNotLeak
- testNotLeakDirectBufferWhenWrapIllegalArgumentException
- testNotLeakWhenWrapIllegalArgumentException
- testDecodeFullHttpRequestWithUrlEncodedBodyWithBrokenHexByte0
- testDecodeFullHttpRequestWithUrlEncodedBodyWithBrokenHexByte1
- testDecodeFullHttpRequestWithUrlEncodedBodyWithInvalidHexNibbleHi
- testDecodeFullHttpRequestWithUrlEncodedBodyWithInvalidHexNibbleLo

They all share the same spirit: allocating a Decoder with a wrong parameter, but request shall be released and as there is no Decoder, we cannot check it but the check on memory leak i the CI/CD get them if they are not correctly handled.

If I removed this, it will certainly failed during leak detection, as it was the case before I added this, and with equal action than to the other standard implementation.

WDYT?

Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:190]: Sounds good.. But please just do it in a separate PR. This will make it easier to review / revert etc if needed.

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:190]: OK, I will remove it, but not sure it will not generate leak detetection errors on CI/CD then. 
And if is so, this PR could not be successful, we shall probably have both but in separate commit if you want. Let see what's happening when removing it.

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java]: As I suspected, removing this is causing leak detection.

This should have been done when the standard decoder was changed, but it wasn't. Now, it's necessary.
So how can we proceed?

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java]: All tests that goes to "Checking for detected leak" have the following similar output:

      Run ./.github/scripts/check_leak.sh build-leak.output
        ./.github/scripts/check_leak.sh build-leak.output
        shell: /usr/bin/bash -e {0}
      Leak detected, please inspect build log
      Error: Process completed with exit code 1.

I repushed the commit with the correction on constructor, as it should be and in line with the other one for Standard decoder. If you have any suggestion, I will go through them.

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:190]: Reference of CI/CD was: https://github.com/netty/netty/pull/11188/checks?check_run_id=2440801961

Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:209]: I factored out this change in its own PR and added a unit test... please have a look https://github.com/netty/netty/pull/11207

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java:209]: OK, I will rebase once merged (review is OK ;-) )

Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostMultipartRequestDecoder.java]: Rebase done, CI/CD on going...

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java [hunk 2]
```diff
@@ -287,13 +287,8 @@ public HttpPostStandardRequestDecoder offer(HttpContent content) {
...
         if (undecodedChunk == null) {
-            undecodedChunk = isLastChunk ?
-                    // Take a slice instead of copying when the first chunk is also the last
-                    // as undecodedChunk.writeBytes will never be called.
-                    buf.retainedSlice() :
-                    // Maybe we should better not copy here for performance reasons but this will need
-                    // more care by the caller to release the content in a correct manner later
-                    // So maybe something to optimize on a later stage.
+            undecodedChunk =
+                    // Since the Handler will release the incoming later on, we need to copy it
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java:291]: @fredericBregier can you explain me why this change ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java (hunk 2 above)
Comment (fredericBregier) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/HttpPostStandardRequestDecoder.java:291]: Several reasons:

- The decoder is not acting in the same way on every chunk : only the first one was `retainedSlice`
- This introduces some bad usages on user's end where the original buffer (from Chunk) were retained and released as soon as possible (the original Chunk), involving memory leak
- It simplifies a lot the mechanism, while it allocates a duplicate ByteBuf, it is as for the next chunks, and performances were not degraded, neither memory pressure (cf Flame on #11175 )

**Comment by normanmaurer** [comment]
@fredericBregier can you please squash and adjust the commit message ?

**Comment by fredericBregier** [comment]
@normanmaurer Done, CI/CD rechecking now ;-)

**Comment by normanmaurer** [comment]
@fredericBregier thanks a lot!
